presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java

```diff
@@ -77,6 +83,11 @@
     private final boolean useNetworkTopology;
     private final Duration nodeMapRefreshInterval;
 
+    @GuardedBy("this")
+    private List<HostAddress> sortedAllNodeList;
```

Comment:
I think we should reuse existing nodeMap rather having a separate list. We should modify the existing nodeMap to a <NodeMap, State>. Based on the configuration, we can decide to keep disabled node or not in the map.
We would also need to have a TTL, if an entry in the map is updated for a certain period, we can remove the entry.
I think this would simplify the implementation a lot.